inference-engine/src/gna_plugin/gna_graph_compiler.cpp

```diff
@@ -199,37 +201,157 @@ void  GNAGraphCompiler::ConstPrimitive(InferenceEngine::CNNLayerPtr constLayer)
+        total_conv_kernel_size *= in_channels;
+        single_conv_kernel_size *= in_channels;
+    }
+    auto actual_kernel_size = details::product(convolution._weights->getTensorDesc().getDims());
+    if (total_conv_kernel_size != actual_kernel_size) {
+        THROW_GNA_LAYER_EXCEPTION(layer) << "weights size mismatch with kernel size";
+    }
 
     // padding of convolution kernel to be multiply of 8
-    uint32_t num_conv_kernel_padding = ALIGN(convolution._kernel_x * num_feature_map_columns, 8)
-        - convolution._kernel_x * num_feature_map_columns;
+    uint32_t num_conv_kernel_padding = ALIGN(single_conv_kernel_size, 8) - single_conv_kernel_size;
+    if (num_conv_kernel_padding == 0) {
+        gnalog() << LAYER_NAME(layer) << "Kernel is aligned \n";
+    } else {
+        gnalog() << LAYER_NAME(layer) << "Kernel padding is " << num_conv_kernel_padding << "\n";
+    }
+
+    // have to pad input to let last kernel meets it's corresponding input
+    // uint32_t num_inputs = num_feature_map_columns * num_feature_map_rows + num_conv_kernel_padding;
+    uint32_t num_inputs = in_width * in_height * in_channels;
+    uint32_t num_input_padding = ALIGN(num_inputs, 8) - num_inputs;
+
+    //  convert to 2D and set GNA input feature map size
+    uint32_t num_feature_map_columns = in_channels * convolution._stride_x * convolution._stride_y;
+    uint32_t num_feature_map_rows = (in_channels * in_height * in_width) / num_feature_map_columns;
+    //if (convolution._stride_x > 1) {
+    //    num_feature_map_rows = (in_channels * in_height * in_width) / convolution._stride_x;
+    //    num_feature_map_columns = convolution._stride_x;
+    //} else if (convolution._stride_y > 1) {
+    //    num_feature_map_rows = (in_channels * in_height * in_width) / convolution._stride_y;
+    //    num_feature_map_columns = convolution._stride_y;
+    //} else {
+    //    num_feature_map_rows = in_channels * in_height * in_width; // (in_height > 1) ? in_height : in_width;
+    //    num_feature_map_columns = 1; // (in_channels > 1) ? in_channels : ((convolution._kernel_x > 1) ? convolution._kernel_x : convolution._kernel_y);
+    //}
+
+    uint32_t num_filters = convolution._out_depth;
+    uint32_t num_filter_coefficients = single_conv_kernel_size + num_conv_kernel_padding;
+    uint32_t num_filter_rows = (num_filter_coefficients - num_conv_kernel_padding) / num_feature_map_columns;
+    uint32_t num_columns_in = num_inputs + num_input_padding;
+    uint32_t num_columns_out = (((num_inputs - single_conv_kernel_size) / num_feature_map_columns) + 1) * convolution._out_depth;
+    //(((num_inputs + num_input_padding - num_filter_coefficients) / num_feature_map_columns) + 1) * convolution._out_depth;
+   // uint32_t num_columns_out_unpadded = (((num_inputs - single_conv_kernel_size) / num_feature_map_columns) + 1) * convolution._out_depth;
```

Comment:
This should not be commented out.  Doing so introduces a bug when the kernel padding exceeds the input padding.  Some outputs will be missing.